File: applications/rtkprojections/rtkprojections.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkprojections_ggo.h"
#include "rtkGgoFunctions.h"
#include "rtkMacro.h"

#include <itkImageFileWriter.h>
#include <itkImageIOFactory.h>
#include <itkVectorImage.h>

#include <iostream>

int
main(int argc, char * argv[])
{
  GGO(rtkprojections, args_info);

  constexpr unsigned int Dimension = 3;
  bool                   inputIsVectorImage = false;
  TRY_AND_EXIT_ON_ITK_EXCEPTION({
    const auto fileNames = rtk::GetProjectionsFileNamesFromGgo(args_info);
    if (!args_info.component_given && !fileNames.empty())
    {
      auto imageIO =
        itk::ImageIOFactory::CreateImageIO(fileNames.front().c_str(), itk::ImageIOFactory::IOFileModeEnum::ReadMode);
      if (imageIO.IsNotNull())
      {
        imageIO->SetFileName(fileNames.front());
        imageIO->ReadImageInformation();
        inputIsVectorImage = imageIO->GetPixelType() == itk::IOPixelEnum::VECTOR;
      }
    }
  })
  if (inputIsVectorImage)
  {
    using OutputImageType = itk::VectorImage<float, Dimension>;
    if (args_info.poisson_given || args_info.gaussian_given)
    {
      std::cerr << "Noise addition is not supported for vector projections" << std::endl;
      return EXIT_FAILURE;
    }

    using ReaderType = rtk::ProjectionsReader<OutputImageType>;
    auto reader = ReaderType::New();
    rtk::SetProjectionsReaderFromGgo<ReaderType, args_info_rtkprojections>(reader, args_info);

    auto writer = itk::ImageFileWriter<OutputImageType>::New();
    writer->SetFileName(args_info.output_arg);
    writer->SetInput(reader->GetOutput());
    TRY_AND_EXIT_ON_ITK_EXCEPTION(writer->UpdateOutputInformation())
    writer->SetNumberOfStreamDivisions(1 + reader->GetOutput()->GetLargestPossibleRegion().GetNumberOfPixels() /
                                             (1024 * 1024 * 4));

    TRY_AND_EXIT_ON_ITK_EXCEPTION(writer->Update())
  }
  else
  {
    using OutputImageType = itk::Image<float, Dimension>;

    using ReaderType = rtk::ProjectionsReader<OutputImageType>;
    auto reader = ReaderType::New();
    rtk::SetProjectionsReaderFromGgo<ReaderType, args_info_rtkprojections>(reader, args_info);

    OutputImageType::Pointer output =
      rtk::AddNoiseFromGgo<OutputImageType, args_info_rtkprojections>(reader->GetOutput(), args_info);

    auto writer = itk::ImageFileWriter<OutputImageType>::New();
    writer->SetFileName(args_info.output_arg);
    writer->SetInput(output);
    TRY_AND_EXIT_ON_ITK_EXCEPTION(writer->UpdateOutputInformation())
    writer->SetNumberOfStreamDivisions(1 + reader->GetOutput()->GetLargestPossibleRegion().GetNumberOfPixels() /
                                             (1024 * 1024 * 4));

    TRY_AND_EXIT_ON_ITK_EXCEPTION(writer->Update())
  }

  return EXIT_SUCCESS;
}

File: applications/rtkprojections/rtkprojections.py
import argparse
import itk
from itk import RTK as rtk


def build_parser():
    parser = rtk.RTKArgumentParser(
        description="Reads raw projection images, converts them to attenuation and stacks them into a single output image file"
    )

    parser.add_argument(
        "--output", "-o", help="Output file name", type=str, required=True
    )
    rtk.add_rtkinputprojections_group(parser)
    rtk.add_rtknoise_group(parser)

    # Parse the command line arguments
    return parser


def process(args_info: argparse.Namespace):

    OutputPixelType = itk.F
    Dimension = 3
    fileNames = rtk.GetProjectionsFileNamesFromArgParse(args_info)

    inputIsVectorImage = False
    if args_info.component is None and fileNames:
        imageio = itk.ImageIOFactory.CreateImageIO(
            fileNames[0], itk.CommonEnums.IOFileMode_ReadMode
        )
        if imageio is not None:
            imageio.SetFileName(fileNames[0])
            imageio.ReadImageInformation()
            inputIsVectorImage = (
                imageio.GetPixelType() == itk.CommonEnums.IOPixel_VECTOR
            )

    if inputIsVectorImage:
        if args_info.poisson is not None or args_info.gaussian is not None:
            raise RuntimeError("Noise addition is not supported for vector projections")

        OutputImageType = itk.VectorImage[OutputPixelType, Dimension]
        reader = rtk.ProjectionsReader[OutputImageType].New()
        rtk.SetProjectionsReaderFromArgParse(reader, args_info)
        if args_info.verbose:
            print(f"Reading projections...")

        writer = itk.ImageFileWriter[OutputImageType].New()
        writer.SetFileName(args_info.output)
        writer.SetInput(reader.GetOutput())
        if args_info.verbose:
            print(f"Writing output to: {args_info.output}")
        writer.Update()
    else:
        OutputImageType = itk.Image[OutputPixelType, Dimension]

        # Projections reader
        reader = rtk.ProjectionsReader[OutputImageType].New()
        rtk.SetProjectionsReaderFromArgParse(reader, args_info)
        if args_info.verbose:
            print(f"Reading projections...")

        # Add noise
        output = rtk.AddNoiseFromArgParse(reader.GetOutput(), args_info)

        # Write
        writer = itk.ImageFileWriter[OutputImageType].New()
        writer.SetFileName(args_info.output)
        writer.SetInput(output)
        if args_info.verbose:
            print(f"Writing output to: {args_info.output}")
        writer.Update()


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()
